subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c

```diff
@@ -0,0 +1,612 @@
+		return err;
+	}
+
+	/* This module loads settings for all application modules */
+	err = settings_load();
+	if (err) {
+		LOG_ERR("Cannot load settings");
+		return err;
+	}
+
+	return err;
+}
+
+static void dps_reg_status_work_fn(struct k_work *work)
+{
+	int err;
+	struct mqtt_publish_param param = {
+		.message.payload.data = NULL,
+		.message.payload.len = 0,
+		.message.topic.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.message_id = k_uptime_get_32(),
+	};
+	static uint32_t retry_count;
+	struct dps_reg_status *reg_status =
+		CONTAINER_OF(work, struct dps_reg_status, poll_work);
+	char topic[DPS_TOPIC_OPERATION_ID_MAX_LEN];
+	size_t topic_len;
+
+	if (dps_reg_status.state != DPS_STATE_REGISTERING) {
+		return;
+	}
+
+	retry_count++;
+
+	if ((retry_count <= DPS_REG_STATUS_UPDATE_MAX_RETRIES) &&
+	    (reg_status->state == DPS_STATE_REGISTERING)) {
+		k_delayed_work_submit(
+			(struct k_delayed_work *)&reg_status->poll_work,
+			K_SECONDS(reg_status->retry));
+	} else if (retry_count > DPS_REG_STATUS_UPDATE_MAX_RETRIES) {
+		LOG_ERR("DPS retry max count (%d) exceeded",
+			DPS_REG_STATUS_UPDATE_MAX_RETRIES);
+		reg_status->state = DPS_STATE_FAILED;
+		cb_handler(reg_status->state);
+
+		return;
+	}
+
+	topic_len = snprintk(topic, sizeof(topic),
+			     DPS_TOPIC_REG_STATUS,
+			     dps_reg_status.reg_id,
+			     dps_reg_status.operation_id);
+
+	param.message.topic.topic.utf8 = topic;
+	param.message.topic.topic.size = topic_len;
+
+	LOG_DBG("Requesting DPS status update");
+
+	err = mqtt_publish(mqtt_client, &param);
+	if (err) {
+		LOG_ERR("Failed to send status request, err: %d", err);
+		return;
+	}
+}
+
+static int dps_get_operation_id(char *json)
+{
+	cJSON *root_obj, *op_id_obj;
+	char *op_id_str;
+	int err = 0;
+
+	root_obj = cJSON_Parse(json);
+	if (root_obj == NULL) {
+		LOG_DBG("[%s:%d] Unable to parse input", __func__, __LINE__);
+		return -ENOMEM;
+	}
+
+	op_id_obj = cJSON_GetObjectItem(root_obj, "operationId");
+	if (op_id_obj == NULL) {
+		LOG_ERR("Could not find operationId object");
+		err = -ENOMEM;
+		goto exit;
+	}
+
+	op_id_str = cJSON_GetStringValue(op_id_obj);
+	if (op_id_str == NULL) {
+		LOG_ERR("Failed to get operation ID string");
+		err = -ENOMEM;
+		goto exit;
+	}
+
+	strncpy(dps_reg_status.operation_id, op_id_str,
+		sizeof(dps_reg_status.operation_id));
+
+exit:
+	cJSON_Delete(root_obj);
+
+	return err;
+}
+
+static int dps_check_reg_success(char *json)
+{
+	cJSON *root_obj, *reg_status_obj, *status_item, *assigned_hub_item;
+	char *status_str, *assigned_hub_str;
+	int err = 0;
+
+	root_obj = cJSON_Parse(json);
+	if (root_obj == NULL) {
+		LOG_DBG("[%s:%d] Unable to parse input", __func__, __LINE__);
+		return -ENOMEM;
+	}
+
+	reg_status_obj =
+		cJSON_GetObjectItem(root_obj, "registrationState");
+	if (reg_status_obj == NULL) {
+		LOG_ERR("Did not find registrationState object");
+		err = -ENOMEM;
+		goto exit;
+	}
+
+	status_item =
+		cJSON_GetObjectItem(reg_status_obj, "status");
+	if (reg_status_obj == NULL) {
```

Comment:
Here the return value should be checked?

Reply:
Good catch, fixed